eth/downloader/downloader.go

```diff
@@ -1281,6 +1343,24 @@ func (d *Downloader) processHeaders(origin uint64, td *big.Int) error {
 					if chunkHeaders[len(chunkHeaders)-1].Number.Uint64()+uint64(fsHeaderForceVerify) > pivot {
 						frequency = 1
 					}
+					// Although the received headers might be all valid, a legacy
+					// PoW/PoA sync must not accept post-merge headers. Make sure
+					// that any transition is rejected at this point.
+					if !beaconMode && ttd != nil {
+						ptd := d.lightchain.GetTd(chunkHeaders[0].ParentHash, chunkHeaders[0].Number.Uint64()-1)
+						if ptd == nil {
+							// This should never really happen, but handle gracefully for now
+							log.Error("Failed to retrieve parent header TD", "number", chunkHeaders[0].Number.Uint64()-1, "hash", chunkHeaders[0].ParentHash)
+							return fmt.Errorf("%w: parent TD missing", errInvalidChain)
+						}
+						for _, header := range chunkHeaders {
+							ptd = new(big.Int).Add(ptd, header.Difficulty)
```

Comment:
If a malicious peer feeds us 1000 legit headers, then at header 1001 ships a header lying about an enormous difficulty -- do we detect that or do we return `ErrMergeTransition` -- and if so, what are the sideeffects?

Reply:
The fault is real in that a fake TD could cause the ErrMergeTransition to be hit. That terminates sync, though does not have other side effects. The issue is that a malicious peer remains connected, so they could spin-loop the same attack, hanging us.

If we check diff progression, that would limit the attack surface to 1-2 blocks right before the merge, at which point our beacon client would take over soon and make the whole thing moot. In addition, this can only happen during initial snap sync. The full sync does is a bit further down and handles TTD exceeding a bit differently.

So, I agree that a TD progression check should be able to avoid the issue in all practical cases.

Reply:
Huh, this isn't that trivial, since the difficulty calculation changes across consensus engines in a non trivial way. I think it will be safer to reproduce the same code that's in full block import (split the range into pre/post merge and import the former).

Reply:
Done like the block importing.